pkg/auth/oauth/external/handler.go

```diff
@@ -145,7 +145,7 @@ func DefaultState() State {
 func (defaultState) Generate(w http.ResponseWriter, req *http.Request) (string, error) {
 	state := url.Values{
 		"csrf": {"..."}, // TODO: get csrf
-		"then": {req.URL.String()},
+		"then": {url.QueryEscape(req.URL.String())},
```

Comment:
Can you make sure keycloak is not altering the state passed to it? For example, the state sent to Google in step 2 is echoed back exactly in step 4. Seems like keycloak might be doing an extra decode, not treating state as an opaque value